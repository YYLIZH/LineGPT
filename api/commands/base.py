import typing
from abc import ABC

from api.utils.configs import LANGUAGE


class Command(ABC):
    usage_en = ""
    usage_zh = ""

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        self.args = args

    def execute(self, **kwargs):
        raise NotImplementedError

    def print_usage(cls):
        return getattr(cls, f"usage_{LANGUAGE}")
