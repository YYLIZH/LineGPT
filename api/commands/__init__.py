from collections import namedtuple
from typing import Dict
from api.utils.configs import LANGUAGE

CommandInfo = namedtuple(
    "CommandInfo",
    "module_path, class_name, summary_en, summary_zh_TW",
)
commands_info: Dict[str, CommandInfo] = {
    "gpt": CommandInfo(
        "api.commands.gpt", "GptCommand", "GPT commands", "GPT 指令"
    ),
    "weather": CommandInfo(
        "api.commands.weather",
        "WeatherCommand",
        "Check the weather",
        "查看天氣",
    ),
    "settle": CommandInfo(
        "api.commands.settle",
        "SettleCommand",
        "Settle the expenses among a group",
        "多人分帳",
    ),
    "eat": CommandInfo(
        "api.commands.eat",
        "EatCommand",
        "What to eat now",
        "現在吃什麼",
    ),
}


def print_usage():
    heading = (
        "嗨！我是LineGPT，很高興為您服務。以下是我的使用說明：\n\n"
        "@LineGPT <指令>\n"
        "可用指令\n"
    )
    footer = "\n您可以使用'@LineGPT <指令> help'來查看更多資訊"
    if LANGUAGE == "en":
        heading = (
            "Hi! I am LineGPT. It's my pleasure to help you. Here is the usage:\n\n"
            "@LineGPT <Command>\n"
            "Available commands\n"
        )
        footer = "\nYou can use '@LineGPT <Command> help' to get more information about each command."

    res = heading
    for key, value in commands_info.items():
        summary = getattr(value, f"summary_{LANGUAGE}")
        line = f"{key:<15}{summary}\n"
        res += line
    res += footer
    return res
