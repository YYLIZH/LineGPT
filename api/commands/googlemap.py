from datetime import datetime, timedelta
from typing import Literal


class GoogleMapSession:
    timeout = 180

    def __init__(self):
        self.last_update_time = datetime.now() - timedelta(days=1)
        self.app = None

    def is_expired(self) -> bool:
        expired_time = self.last_update_time + timedelta(
            seconds=self.timeout
        )
        if datetime.now() > expired_time:
            return True
        return False

    def reset(self):
        self.app = None
        self.last_update_time = datetime.now() - timedelta(days=1)

    def set_app(self, app: Literal["eat", "toilet"]):
        self.last_update_time = datetime.now()
        self.app = app


GOOGLE_MAP_SESSION = GoogleMapSession()
