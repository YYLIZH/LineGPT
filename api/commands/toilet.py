import csv
import math
import re
import textwrap
from datetime import datetime, timedelta
from enum import Enum
from functools import cache
from pathlib import Path

from api.utils.configs import LANGUAGE


class GoogleMapSession:
    timeout = 5

    def __init__(self):
        self.last_update_time = datetime.now() - timedelta(days=1)

    def update_time(self):
        self.last_update_time = datetime.now()

    def is_expired(self) -> bool:
        expired_time = self.last_update_time + timedelta(
            minutes=self.timeout
        )
        if datetime.now() > expired_time:
            return True
        return False

    def set_expired(self):
        self.last_update_time = datetime.now() - timedelta(days=1)


GOOGLE_MAP_SESSION = GoogleMapSession()


class MessageEN(str, Enum):
    START_REPLY = "Please share your location to me"
    STOP_REPLY = "Stop session. Please start again if you want to search toilet"
    NO_RESULT = "No result in nearby location."


class MessageZHTW(str, Enum):
    START_REPLY = "請分享你的位置資訊給我"
    STOP_REPLY = "關閉對話。如果您想繼續搜尋，請重新開始對話"
    NO_RESULT = "找不到結果"


MESSAGE = MessageZHTW if LANGUAGE == "zh_TW" else MessageEN


@cache
def read_toilet_data() -> list[dict]:
    """Read toilet data"""
    toilet_data = []
    with (Path(__file__).parent / "data" / "toilet.csv").open(
        "r", encoding="utf-8", newline=""
    ) as filep:
        rows = csv.DictReader(filep)

        for row in rows:
            # remove \u3000 space
            for key in row:
                row[key]=row[key].replace("\u3000","")
            toilet_data.append(row)

    return toilet_data


def get_toilets(
    latitude: str, longitude: str
) -> list[tuple[float, dict]]:
    toilet_data = read_toilet_data()
    distances_out = []

    for data in toilet_data:
        toilet, distance = curr2toilet((latitude, longitude), data)
        distances_out.append((distance, toilet))

    distances_out.sort(key=lambda item: item[0])
    return distances_out[:10]


def create_map_link(latitude: str, longitude: str) -> str:
    map_link = f"https://www.google.com/maps/search/?api=1&query={latitude}%2C{longitude}"
    return map_link


def calculate_distance(
    lat1: float, lon1: float, lat2: float, lon2: float
) -> float:
    """Use Haversine Formula

    Returns:
        The return value unit is km.

    """
    # 地球半徑（公里）
    R = 6371.0

    # 將經緯度轉換為弧度
    lat1_rad = math.radians(lat1)
    lon1_rad = math.radians(lon1)
    lat2_rad = math.radians(lat2)
    lon2_rad = math.radians(lon2)

    # 經度和緯度的差值
    dlat = lat2_rad - lat1_rad
    dlon = lon2_rad - lon1_rad

    # 哈弗辛公式
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(lat1_rad)
        * math.cos(lat2_rad)
        * math.sin(dlon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    # 距離（公里）
    distance = R * c

    return distance


def curr2toilet(
    curr_location: tuple[float, float], toilet: dict
) -> tuple[dict, float]:
    distance = calculate_distance(
        float(curr_location[0]),
        float(curr_location[1]),
        float(toilet["latitude"]),
        float(toilet["longitude"]),
    )
    return toilet, distance


def get_place_detail(toilet_calculation: tuple[float, dict]):
    distance, toilet_info = toilet_calculation
    return {
        "name": toilet_info["name"],
        "address": toilet_info["address"],
        "distance": f"{int(distance*1000)} m",
        "map_url": create_map_link(
            toilet_info["latitude"], toilet_info["longitude"]
        ),
    }


def generate_card(
    name: str,
    address: str,
    distance: float,
    map_url: str,
) -> dict:
    card = {
        "type": "bubble",
        "size": "micro",
        "body": {
            "type": "box",
            "layout": "vertical",
            "contents": [
                {
                    "type": "text",
                    "text": str(name),
                    "weight": "bold",
                    "size": "sm",
                    "wrap": True,
                },
                {
                    "type": "box",
                    "layout": "baseline",
                    "contents": [],
                },
                {
                    "type": "box",
                    "layout": "vertical",
                    "contents": [
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(address),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        },
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(distance),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        },
                    ],
                },
            ],
            "spacing": "sm",
            "paddingAll": "13px",
        },
        "footer": {
            "type": "box",
            "layout": "horizontal",
            "contents": [
                {
                    "type": "button",
                    "action": {
                        "type": "uri",
                        "label": "Go",
                        "uri": map_url,
                    },
                    "style": "primary",
                }
            ],
        },
    }
    return card


def generate_flex_message(place_details: list[dict]) -> dict:
    flex_message_template = {
        "type": "carousel",
        "contents": [],
    }

    for detail in place_details:
        flex_message_template["contents"].append(
            generate_card(**detail)
        )

    return flex_message_template


def where_to_pee(latitude: str, longitude: str) -> dict:
    toilets = get_toilets(latitude, longitude)
    if len(toilets) == 0:
        return MESSAGE.NO_RESULT.value

    # Format the message
    place_details = []
    for toilet in toilets:
        place_detail = get_place_detail(toilet_calculation=toilet)
        place_details.append(place_detail)

    flex_message_dict = generate_flex_message(place_details)
    return flex_message_dict


def print_help() -> str:
    usage_en = textwrap.dedent(
        """
        * Start a new session to find toilet nearby.

        Example:
        @LineGPT toilet start

        * Stop a current session to use other location function.

        Example:
        @LineGPT toilet stop
        """
    )

    usage_zh_TW = textwrap.dedent(
        """
        * 開啟對話來尋找附近的廁所

        Example: 
        @LineGPT toilet start

        * 關閉目前對話來使用其他定位功能

        Example:
        @LineGPT toilet stop
        """
    )
    if LANGUAGE == "zh_TW":
        return usage_zh_TW
    return usage_en


def handle_message(message: str) -> str:
    if "help" in message:
        return print_help()

    if mrx := re.search(r"^toilet\s+(\w+)", message):
        match mrx.group(1):
            case "start":
                GOOGLE_MAP_SESSION.update_time()
                return MESSAGE.START_REPLY.value
            case "stop":
                GOOGLE_MAP_SESSION.set_expired()
                return MESSAGE.STOP_REPLY.value
            case _:
                return print_help()

    return print_help()
