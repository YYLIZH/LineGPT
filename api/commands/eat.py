from __future__ import annotations

import math
import textwrap
from datetime import datetime, timedelta
from enum import Enum

import requests

from api.commands.base import Command
from api.utils.configs import GOOGLE_API_KEY, LANGUAGE


class MessageEN(str, Enum):
    REPLY = "Please share your location to me"
    NO_RESULT = "No result in nearby location."


class MessageZHTW(str, Enum):
    REPLY = "請分享你的位置資訊給我"
    NO_RESULT = "找不到結果"


MESSAGE = MessageEN if LANGUAGE == "en" else MessageZHTW


def get_food(latitude: str, longitude: str) -> list[dict]:
    radius = 3000
    place_type = "restaurant"
    location = f"{latitude},{longitude}"

    query_opts = [
        f"location={location}",
        f"radius={radius}",
        f"type={place_type}",
        f"key={GOOGLE_API_KEY}",
        "opennow=true",
    ]
    if LANGUAGE == "zh_TW":
        query_opts.append("language=zh-TW")

    url = f"https://maps.googleapis.com/maps/api/place/nearbysearch/json?" + "&".join(
        query_opts
    )

    try:
        response = requests.get(url)
        data = response.json()
        return sorted(data["results"], key=lambda x: x["rating"], reverse=True)[:10]
    except Exception:
        return []


def create_map_link(latitude: str, longitude: str, place_id: str) -> str:
    map_link = f"https://www.google.com/maps/search/?api=1&query={latitude}%2C{longitude}&query_place_id={place_id}"
    return map_link

def calculate_distance(lat1:float, lon1:float, lat2:float, lon2:float)->str:
    """Use Haversine Formula"""
    # 地球半徑（公里）
    R = 6371.0
    
    # 將經緯度轉換為弧度
    lat1_rad = math.radians(lat1)
    lon1_rad = math.radians(lon1)
    lat2_rad = math.radians(lat2)
    lon2_rad = math.radians(lon2)
    
    # 經度和緯度的差值
    dlat = lat2_rad - lat1_rad
    dlon = lon2_rad - lon1_rad
    
    # 哈弗辛公式
    a = math.sin(dlat / 2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    # 距離（公里）
    distance = R * c

    if distance>1:
        return f"{round(distance,2)} km"

    return f"{round(distance*1000,2)} m"

def get_place_detail(user_lat:str,user_lng:str,place_info: dict) -> dict[str, str]:
    name = str(place_info.get("name", ""))
    rating = place_info.get("rating", 0.0)
    user_ratings_total = place_info.get("user_ratings_total", 0)
    address = str(place_info.get("vicinity", "N/A"))

    latitude = place_info["geometry"]["location"]["lat"]
    longitude = place_info["geometry"]["location"]["lng"]

    return {
        "name": name,
        "address": address,
        "distance":calculate_distance(float(user_lat),float(user_lng),latitude,longitude),
        "rating": rating,
        "user_ratings_total": user_ratings_total,
        "map_url": create_map_link(latitude, longitude, place_info["place_id"]),
    }


def generate_card(
    name: str, address:str,distance:float,rating: float, user_ratings_total: int, map_url: str
) -> dict:
    card = {
        "type": "bubble",
        "size": "micro",
        "body": {
            "type": "box",
            "layout": "vertical",
            "contents": [
                {
                    "type": "text",
                    "text": str(name),
                    "weight": "bold",
                    "size": "sm",
                    "wrap": True,
                },
                {
                    "type": "box",
                    "layout": "baseline",
                    "contents": [],
                },
                {
                    "type": "box",
                    "layout": "baseline",
                    "contents": [
                        {
                            "type": "text",
                            "text": f"{user_ratings_total} votes",
                            "size": "sm",
                        }
                    ],
                },
                {
                    "type": "box",
                    "layout": "vertical",
                    "contents": [
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(address),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        },
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(distance),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        }
                    ],
                },
            ],
            "spacing": "sm",
            "paddingAll": "13px",
        },
        "footer": {
            "type": "box",
            "layout": "horizontal",
            "contents": [
                {
                    "type": "button",
                    "action": {
                        "type": "uri",
                        "label": "Go",
                        "uri": map_url,
                    },
                    "style": "primary",
                }
            ],
        },
    }

    # calculate rating box content
    rating_box_content = []
    rating_floor = math.ceil(rating)
    num_golden = rating_floor
    num_empty = 5 - num_golden
    for _ in range(num_golden):
        rating_box_content.append(
            {
                "type": "icon",
                "size": "xs",
                "url": "https://developers-resource.landpress.line.me/fx/img/review_gold_star_28.png",
            }
        )
    for _ in range(num_empty):
        rating_box_content.append(
            {
                "type": "icon",
                "size": "xs",
                "url": "https://developers-resource.landpress.line.me/fx/img/review_gray_star_28.png",
            }
        )

    rating_box_content.append(
        {
            "type": "text",
            "text": str(rating),
            "size": "xs",
            "color": "#8c8c8c",
            "margin": "md",
            "flex": 0,
        }
    )
    card["body"]["contents"][1]["contents"] = rating_box_content
    return card


def generate_flex_message(place_details: list[dict]) -> dict:
    flex_message_template = {
        "type": "carousel",
        "contents": [],
    }

    for detail in place_details:
        flex_message_template["contents"].append(generate_card(**detail))

    return flex_message_template


def what_to_eat(latitude: str, longitude: str) -> dict:
    restaurants = get_food(latitude, longitude)
    if len(restaurants) == 0:
        return MESSAGE.NO_RESULT.value

    # Format the text
    place_details = []
    for restaurant in restaurants:
        place_detail = get_place_detail(latitude,longitude,restaurant)
        place_details.append(place_detail)

    flex_message_dict = generate_flex_message(place_details)

    return flex_message_dict


class GoogleMapSession:
    timeout = 5

    def __init__(self):
        self.last_update_time = datetime.now() - timedelta(days=1)

    def update_time(self):
        self.last_update_time = datetime.now()

    def is_expired(self) -> bool:
        expired_time = self.last_update_time + timedelta(minutes=self.timeout)
        if datetime.now() > expired_time:
            return True
        return False


class EatCommand(Command):
    usage_en = textwrap.dedent(
        """
        * Find opening restaurant nearby.

        Example:
        @LineGPT eat
        """
    )

    usage_zh_TW = textwrap.dedent(
        """
        * 尋找附近的餐廳

        Example: 
        @LineGPT eat
        """
    )

    def __init__(self, subcommand=None, args=None):
        super().__init__(subcommand, args)
        self.google_map_session = None

    @classmethod
    def setup(cls, args_msg: str) -> EatCommand:
        return cls(None, args_msg)

    def load(self, google_map_session: GoogleMapSession):
        self.google_map_session = google_map_session

    def execute(self, **kwargs):
        return MESSAGE.REPLY.value
