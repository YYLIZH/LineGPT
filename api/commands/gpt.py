from __future__ import annotations
import datetime
import re
import typing
from enum import Enum
from typing import Dict

import openai

from api.commands.base import Command
from api.utils.configs import (
    LANGUAGE,
    OPENAI_API_KEY,
    OPENAI_FREQUENCY_PENALTY,
    OPENAI_MAX_TOKEN,
    OPENAI_MODEL,
    OPENAI_PRESENCE_PENALTY,
    OPENAI_TEMPERATURE,
    SESSION_EXPIRED,
)
from api.utils.info import Error, Warning


class MessageEN(Enum):
    GREETING = "Start the dialogue session."
    FOUND_SESSION = "Found existing session. You can see existing session's dialog, keep talking, or restart a new session."
    NO_SESSION = "Session does not exist. Please start the session first."
    SESSION_EXPIRED = (
        "Current dialogue session is expired. Please restart the dialogue session."
    )
    RUNTIME_ERROR = "RuntimeError. Please ask again later."
    UNEXPECTED_ERROR = "Unexpected error happened. Restart the session."
    CLOSE = "Close current dialogue session."
    RESTART = "Restart the dialogue session"
    LOG = "Below are the past dialogues:"
    NOT_ALLOW_METHOD = "This method is not allowed."


class MessageZHTW(Enum):
    GREETING = "對話階段已開始"
    FOUND_SESSION = "存在舊有的對話階段，您可以查看紀錄、繼續對話、或是重啟對話階段。"
    NO_SESSION = '尚未開啟對話階段，請先使用"@LineGPT start"開啟對話。'
    SESSION_EXPIRED = '當前對話階段已過期。請先使用"@LineGPT restart"重新開啟對話階段。'
    RUNTIME_ERROR = "RuntimeError. 請稍後再試一次"
    UNEXPECTED_ERROR = "發生未知的錯誤，重啟對話。"
    CLOSE = "關閉對話階段。"
    RESTART = "重新啟動對話階段"
    LOG = "以下是對話紀錄："
    NOT_ALLOW_METHOD = "不存在的指令"


USAGE_EN = """* Start a dialogue session.
@LineGPT gpt start

* Show dialogue
@LineGPT gpt log

* Ask a question
@LineGPT gpt talk <your question> 
Example:
@LineGPT gpt talk What is graphene?

* Close an existing dialogue session
@LineGPT gpt close
"""

USAGE_ZH_TW = """* 開始對話階段
@LineGPT gpt start

* 顯示過往對話紀錄
@LineGPT gpt log

* 提問
@LineGPT gpt talk <your question> 
Example:
@LineGPT gpt talk 請告訴我怎麼變有錢人

* 關閉對話階段，過往紀錄將會被刪除
@LineGPT gpt close
"""

openai.api_key = OPENAI_API_KEY
MESSAGE = MessageEN if LANGUAGE == "en" else MessageZHTW


class DialogueSession:
    def __init__(self) -> None:
        self.dialogue = []
        self.last_update_time = datetime.datetime.now()

    def add_ai_text(self, text: str) -> None:
        self.dialogue.append(f"AI: {text}")
        self.last_update_time = datetime.datetime.now()

    def add_human_text(self, text: str) -> None:
        self.dialogue.append(f"Human: {text}")
        self.last_update_time = datetime.datetime.now()

    def dump_dialogue(self) -> str:
        return "\n".join(self.dialogue)

    def is_expired(self) -> bool:
        time_delta: datetime.timedelta = datetime.datetime.now() - self.last_update_time
        if time_delta.seconds >= int(SESSION_EXPIRED):
            return True
        return False


class GPT:
    def __init__(self) -> None:
        self.dialogue_session = DialogueSession()
        self.model = OPENAI_MODEL
        self.temperature = OPENAI_TEMPERATURE
        self.max_tokens = OPENAI_MAX_TOKEN
        self.frequency_penalty = OPENAI_FREQUENCY_PENALTY
        self.presence_penalty = OPENAI_PRESENCE_PENALTY

    def _talk(self):
        response = openai.Completion.create(
            model=self.model,
            prompt=self.dialogue_session.dump_dialogue(),
            temperature=self.temperature,
            max_tokens=self.max_tokens,
            frequency_penalty=self.frequency_penalty,
            presence_penalty=self.presence_penalty,
        )
        text = response["choices"][0]["text"].strip()
        extracted_text = re.search(r"(\w+: +)?([\s\S]*)", text).group(2)
        extracted_text = extracted_text.replace("AI: ", "")
        self.dialogue_session.add_ai_text(extracted_text)
        return extracted_text

    def talk(self, text):
        if self.dialogue_session is None:
            return Warning(MESSAGE.NO_SESSION.value)
        if self.dialogue_session.is_expired():
            return Warning(MESSAGE.SESSION_EXPIRED.value)

        self.dialogue_session.add_human_text(text)
        try:
            response = self._talk()
        except Exception:
            self.dialogue_session.dialogue.pop()
            return Error(MESSAGE.RUNTIME_ERROR.value)
        return response


class GPTSessions:
    def __init__(self) -> None:
        """Each Group use different gpt."""
        self.sessions: Dict[str, GPT] = {}

    def start(self, group_id: str) -> str:
        """Start a GPT and add into session."""
        if gpt := self.sessions.get(group_id):
            return MESSAGE.FOUND_SESSION.value
        gpt = GPT()
        self.sessions[group_id] = gpt
        return MESSAGE.GREETING.value

    def close(self, group_id: str) -> str:
        self.sessions.pop(group_id)
        return MESSAGE.CLOSE.value

    def restart(self, group_id: str) -> str:
        gpt = GPT()
        self.sessions[group_id] = gpt
        return MESSAGE.GREETING.value

    def talk(self, group_id: str, text: str) -> str:
        if not self.sessions.get(group_id):
            return Warning(MESSAGE.NO_SESSION.value)
        gpt = self.sessions.get(group_id)
        try:
            return gpt.talk(text)
        except Exception:
            self.sessions[group_id] = GPT()
            return Error(MESSAGE.UNEXPECTED_ERROR.value)

    def log(self, group_id: str) -> str:
        gpt = self.sessions.get(group_id)
        return MESSAGE.LOG.value + "\n" + gpt.dialogue_session.dump_dialogue()


class GptCommand(Command):
    usage_en = USAGE_EN
    usage_zh_TW = USAGE_ZH_TW

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        super().__init__(subcommand, args)
        self.gpt_sessions: GPTSessions = None

    def load(self, gpt_sessions: GPTSessions):
        self.gpt_sessions = gpt_sessions

    @classmethod
    def setup(cls, args_msg: str) -> GptCommand:
        mrx = re.search(r"(\w+) *(.*)?", args_msg)
        subcommand, args = mrx.groups()
        return cls(subcommand, args)

    def execute(self, **kwargs):
        id = kwargs["id"]
        try:
            func = getattr(self.gpt_sessions, self.subcommand)
            if self.subcommand == "talk":
                return func(id, self.args)
            return func(id)
        except AttributeError:
            return Error(MESSAGE.NOT_ALLOW_METHOD.value)
