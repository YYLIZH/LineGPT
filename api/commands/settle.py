from __future__ import annotations
import re
import typing

from api.commands.base import Command
from api.utils.configs import LANGUAGE


class SettleCommand(Command):
    usage_en = """* Settle the money transaction
@LineGPT settle
<chart>

Example:
@LineGPT settle
Iron man: 100
Batman: 300
Superman: 240
Spiderman: 0

Here is the result:
Spiderman -> Batman 140.0
Iron man -> Superman 60.0
Spiderman -> Superman 20.0
"""

    usage_zh_TW = """* 多人分帳
@LineGPT settle
<chart>

Example:
@LineGPT settle
大壯: 100
小帥: 300
小美: 240
大黑: 0

以下是分帳結果：
大黑 -> 小帥 140.0
大壯 -> 小美 60.0
大黑 -> 小美 20.0
"""

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        args = args.replace("：", ":")  # Replace chinese full colon
        super().__init__(subcommand, args)
        self.expenses = {}

    @classmethod
    def setup(cls, args_msg: str) -> SettleCommand:
        args = args_msg.lstrip()
        return cls(None, args)

    def execute(self, **kwargs):
        self.parse_expense()
        transactions = self.settle_money()
        heading = "Here is the result:\n" if LANGUAGE == "en" else "以下是分帳結果：\n"

        return heading + "\n".join(
            [
                f"{transaction[0]} -> {transaction[1]} {transaction[2]}"
                for transaction in transactions
            ]
        )

    def parse_expense(self):
        pattern = re.compile(r"(.+): *(\d+)")
        for line in self.args.split("\n"):
            if mrx := pattern.search(line):
                self.expenses[mrx.group(1)] = int(mrx.group(2))

    def settle_money(self):
        """
        Calculates the amount each person owes or is owed after settling expenses.

        Arguments:
        expenses -- a dictionary of the form {person: amount_paid} representing the expenses paid by each person

        Returns:
        A list of tuples representing the money transactions needed to settle the expenses.
        """
        total_expenses = sum(self.expenses.values())
        num_people = len(self.expenses)
        average_expense = total_expenses / num_people
        owed = {
            person: amount_paid - average_expense
            for person, amount_paid in self.expenses.items()
        }

        transactions = []
        while any(owed.values()):
            person1, amount1 = max(owed.items(), key=lambda x: x[1])
            person2, amount2 = min(owed.items(), key=lambda x: x[1])
            amount = min(-amount2, amount1)
            transactions.append((person2, person1, amount))
            owed[person1] -= amount
            owed[person2] += amount

        return transactions
