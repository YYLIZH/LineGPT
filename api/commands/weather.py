from collections import defaultdict
from enum import Enum
from typing import Optional, Tuple
from urllib import parse

import requests
from jinja2 import Template

from api.commands.base import Command
from api.utils.configs import LANGUAGE, WEATHER_TOKEN
from api.utils.info import Error, Warning

REPORT: Template = Template(
    """
{{ startTime }} - {{ endTime }}
{{Wx}}
降雨機率: {{PoP}} %
{{CI}}
最高溫: {{MaxT}} °C
最低溫: {{MinT}} °C""",
    trim_blocks=True,
)


class MessageEN(Enum):
    WEATHER_HEADING = "Weather forecast in 36 hours:"
    UNAVAILABLE_LOCATION = "Unavailable location"
    GET_DATA_FAILED = "Failed to get data"
    CANNOT_WORK_COLD = "It's too cold to work tokday"
    CANNOT_WORK_HOT = "It's too hot to work tokday"
    CANNOT_WORK_RAIN = "It will rain cats and dogs. Better not leave you house."
    REMIND_UMBRELLA = "It may rain today. Bring an umbrella with you."


class MessageZH(Enum):
    WEATHER_HEADING = "未來36小時天氣預報:"
    UNAVAILABLE_LOCATION = "該地區不適用"
    GET_DATA_FAILED = "獲取資料失敗"
    CANNOT_WORK_COLD = "今天太冷，不要去上班比較好，會凍死在路上"
    CANNOT_WORK_HOT = "今天太熱，不要去上班比較好，會熱死在路上"
    CANNOT_WORK_RAIN = "明天可能會下大雨，不要去上班比較好，太危險了"
    REMIND_UMBRELLA = "今天可能會下雨，出門記得帶傘"


MESSAGE = MessageEN if LANGUAGE == "en" else MessageZH


class WeatherCommand(Command):
    available_location = [
        "宜蘭縣",
        "花蓮縣",
        "臺東縣",
        "澎湖縣",
        "金門縣",
        "連江縣",
        "臺北市",
        "新北市",
        "桃園市",
        "臺中市",
        "臺南市",
        "高雄市",
        "基隆市",
        "新竹縣",
        "新竹市",
        "苗栗縣",
        "彰化縣",
        "南投縣",
        "雲林縣",
        "嘉義縣",
        "嘉義市",
        "屏東縣",
    ]
    usage_en = """* Check for the weather in the next 36 hours
@LineGPT weather <location>

Example:
@LineGPT weather 嘉義縣
        """
    usage_zh = """* 查詢未來36小時的天氣預報
@LineGPT weather <地點>

Example:
@LineGPT weather 嘉義縣
        """

    def __init__(
        self, subcommand: Optional[str] = None, args: Optional[str] = None
    ) -> None:
        super().__init__(subcommand, args)
        self.location = self.args

    def execute(self, **kwargs):
        if self.location not in self.available_location:
            return Error(MESSAGE.UNAVAILABLE_LOCATION.value)
        data = requests.get(
            (
                "https://opendata.cwb.gov.tw/api/v1/rest/datastore/F-C0032-001"
                f"?Authorization={WEATHER_TOKEN}"
                "&format=JSON"
                f"&locationName={parse.quote(self.location)}"
            )
        ).json()

        if data["success"] != "true":
            return Warning(MESSAGE.GET_DATA_FAILED.value)

        data_dict = defaultdict(list)
        for el in data["records"]["location"][0]["weatherElement"]:
            for time in el["time"]:
                data_dict[el["elementName"]].append(time["parameter"]["parameterName"])
        for time in data["records"]["location"][0]["weatherElement"][0]["time"]:
            data_dict["startTime"].append(time["startTime"])
            data_dict["endTime"].append(time["endTime"])

        message = []
        if int(min(data_dict["MinT"])) < 15:
            message.append(MESSAGE.CANNOT_WORK_COLD.value)

        if int(max(data_dict["MaxT"])) > 30:
            message.append(MESSAGE.CANNOT_WORK_HOT.value)

        if 100 > int(min(data_dict["PoP"])) > 60:
            message.append(MESSAGE.REMIND_UMBRELLA.value)

        if int(min(data_dict["PoP"])) == 100:
            message.append(MESSAGE.CANNOT_WORK_RAIN.value)

        report_list = []
        for i in range(3):
            report_data = {}
            for key in data_dict:
                report_data[key] = data_dict[key][i]
            report_list.append(REPORT.render(report_data))

        return (
            MESSAGE.WEATHER_HEADING.value
            + "\n"
            + "\n".join(report_list)
            + "\n\n"
            + ",".join(message)
        )


def parse_args(args_msg: str) -> Tuple[str, str]:
    args = args_msg.strip()
    return None, args
