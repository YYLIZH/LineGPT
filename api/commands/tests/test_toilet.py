from datetime import datetime

from api.commands import toilet


def test_get_toilets(snapshot):
    latitude = "24.785057"
    longitude = "121.017385"
    toilets = toilet.get_toilets(latitude, longitude)
    assert snapshot == toilets[0]


class TestGoogleMapSession:
    def test_init(self):
        session = toilet.GoogleMapSession()
        assert (datetime.now() - session.last_update_time).days >= 1

    def test_update_time(self):
        session = toilet.GoogleMapSession()
        session.update_time()
        assert (
            datetime.now() - session.last_update_time
        ).seconds < 10

    def test_is_expired(self):
        session = toilet.GoogleMapSession()
        assert session.is_expired() is True

        session.update_time()
        assert session.is_expired() is False

    def test_set_expired(self):
        session = toilet.GoogleMapSession()
        session.update_time()
        session.set_expired()
        assert session.is_expired() is True


def test_print_help(snapshot):
    help_message = toilet.print_help()
    assert snapshot == help_message


def test_handle_message_toilet():
    msg = toilet.handle_message("toilet start")
    assert msg == toilet.MESSAGE.START_REPLY.value

    msg = toilet.handle_message("toilet stop")
    assert msg == toilet.MESSAGE.STOP_REPLY.value

    msg = toilet.handle_message("toilet help")
    assert msg == toilet.print_help()

    assert toilet.handle_message(
        "toilet help aaa"
    ) == toilet.handle_message("toilet aaa help")
