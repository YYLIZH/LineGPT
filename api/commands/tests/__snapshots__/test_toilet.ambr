# serializer version: 1
# name: test_get_toilets
  tuple(
    0.018799897788212956,
    dict({
      'address': '臺北市信義區西村里信義路五段7號',
      'latitude': '25.03352400',
      'longitude': '121.56460400',
      'name': '台北101觀景台91樓男',
    }),
  )
# ---
# name: test_print_help
  '''
  
  * 開啟對話來尋找附近的廁所
  
  Example: 
  @LineGPT toilet start
  
  * 關閉目前對話來使用其他定位功能
  
  Example:
  @LineGPT toilet stop
  
  '''
# ---
